### PR TITLE
Issue812 merge lideas

### LINKED ISSUE #812: Merge LIDEAS into IDEAS
I'll propose a pull request including some unit tests

### PR DESCRIPTION
**Pull request by damienpicard**
This pull requests merges LIDEAS into IDEAS and add some validation examples to be included in the unit tests suits.

Unit tests still need to be ran.

This closes #812

### CODE REVIEW

**Comment by Mathadon** [comment]
@damienpicard should the SSM error of `IDEAS.LIDEAS.Validation.Case900ValidationLinear` not be exactly zero?

**Comment by Mathadon** [comment]
@damienpicard the examples and results look good to me!

There are some things that we should look into though:
 - [ ] Should the SSM error of IDEAS.LIDEAS.Validation.Case900ValidationLinear not be exactly zero?
 - [ ] The documentation contains a lot of "&QUOT;"s etc. This will produce a lot of html errors once we enable the html check in travis. I hope you can fix it then, or now =)
 - [ ] Similarly, we should have models that produce no warnings. This implies that stuff like `Redeclaration of non-replaceable requires type equivalence
But missing component indexWindow.` and `Connect argument was not one of the valid forms, since sim is not a connector.` should probably be sorted out.
 - [ ] According to IBPSA guidelines each model should have a comment, each component in the model should have a comment and there should be a documentation section and revision history for each model. I think we should start adhering to this to avoid a backlog. At some point these requirements will most likely also be checked in the unit tests.
 - [ ] Some of your .mos scripts do not contain plot commands. This will cause them to fail the unit tests. For the linearisation examples it would make sense to read the .mat files that are written and plot the first line of the state space matrices such that these results are also unit tested.
 - [ ] Why is the `SimInfoManager` replaceable?

I also pushed some minor changes.

**Comment by Mathadon** [comment]
@damienpicard I merged issue819_html such that travis is now run

**Comment by damienpicard** [comment]
- Should  the SSM error of IDEAS.LIDEAS.Validation.Case900ValidationLinear not be exactly zero?
--> not all linear flag were set to linear, so it is ok. The model `IDEAS.LIDEAS.Validation.ZoneWithInputsValidationLinear` shows that the linearization error is smaal (10^-05).
- Please fix html errors (see https://travis-ci.org/open-ideas/IDEAS/jobs/380058694)
--> should be done
- Similarly, we should have models that produce no warnings. This implies that stuff like Redeclaration of non-replaceable requires type equivalence But missing component indexWindow. and Connect argument was not one of the valid forms, since sim is not a connector. should probably be sorted out.
--> do you have time to look at that?
- According to IBPSA guidelines each model should have a comment, each component in the model should have a comment and there should be a documentation section and revision history for each model. I think we should start adhering to this to avoid a backlog. At some point these requirements will most likely also be checked in the unit tests.
--> done
- Some of your .mos scripts do not contain plot commands. This will cause them to fail the unit tests. For the linearisation examples it would make sense to read the .mat files that are written and plot the first line of the state space matrices such that these results are also unit tested.
--> only the validation models are meant to be part of the unit test suits. All scripts included in `Resources/Script/Dymola/...` have plot commands. The linearization scripts (when not included in the unit test script) are in the `IDEAS/Scripts` folder.
--> Do you really think that plotting the matrices element is an added value? If so, I will just plot some A values from the state space model which is included in the validation model (instead of manually read the .mat file)
- Why is the SimInfoManager replaceable?
--> because there is an example model which shows how to use a custom weather file (`IDEAS.LIDEAS.Examples.ZoneWithInputsCreateOutputs_customWeather`). This model uses it's own SimInfoManager (`IDEAS.LIDEAS.Examples.BaseClasses.SimInfoManagerInputs`).

**Comment by damienpicard** [comment]
@Mathadon I have not made the time yet to include all documentation from D3.1. However, this is not urgent and the pull request can be merged with it.

**Comment by Mathadon** [comment]
@damienpicard , looks good!

**Comment by Mathadon** [comment]
@damienpicard can you fix the errors in the unit tests?

**Comment by damienpicard** [comment]
Model should run with DryAir. In order to have only temperature states, dynBal should be parameterized: `dynBal(T(stateSelect=StateSelect.prefer))`.

An assert should also be added checking that `Medium.nXi < 2` and `Medium.nC = 0`

**Comment by damienpicard** [comment]
@Mathadon I'm stuck at the warnings from `IDEAS.LIDEAS.Components.LinWindow`. I don't see what's wrong with the connect statement.

**Comment by Mathadon** [comment]
@damienpicard unit tests pass. Is this ready for review?

File: IDEAS/Buildings/Components/RectangularZoneTemplate.mo [hunk 1]
```diff
@@ -333,8 +334,9 @@ protected
     fraType(present=fraTypA.present,
             U_value=fraTypA.U_value),
     linExtRad=linExtRadWin) if
-       hasWinA
+       hasWinA constrainedby Window
```
Comment (Mathadon) [inline IDEAS/Buildings/Components/RectangularZoneTemplate.mo]: the parameters
```
      inc=IDEAS.Types.Tilt.Wall,
    glazing(
      nLay=glazingB.nLay,
      mats=glazingB.mats,
      SwAbs=glazingB.SwAbs,
      SwTrans=glazingB.SwTrans,
      SwAbsDif=glazingB.SwAbsDif,
      SwTransDif=glazingB.SwTransDif,
      U_value=glazingB.U_value,
      g_value=glazingB.g_value),
    A=A_winB,
    frac=fraB,
    azi=aziA + Modelica.Constants.pi/2,
    T_start=T_start,
    linIntCon_a=linIntCon,
    dT_nominal_a=dT_nominal_win,
    linExtCon=linExtCon,
    windowDynamicsType=windowDynamicsType,
```
etc have to be in the `constrainedby` clause when making a component replaceable. Otherwise the parameter values can be 'forgotten' when redeclaring a component. This depends on the Modelica tool that is used, but better be safe than sorry!

File: IDEAS/LIDEAS/Components/BaseClasses/Inputs.mo [hunk 2]
```diff
@@ -0,0 +1,9 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+expandable connector Inputs
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/Inputs.mo]: Can we drop the `expandable` here? If so, then I would prefer that to avoid wrong use and to allow the Modelica tools to have easier error checking.

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo [hunk 3]
```diff
@@ -0,0 +1,270 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+model LinWindowResponse
+  "Shortwave window respones implementation for linearisation"
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: 'response'

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+  parameter Boolean createOutputs = false
+    "Set to true for enabling linearization inputs/outputs";
+
+  final parameter Integer[nLay] columns=if (nLay == 1) then {2} else integer(
+      linspace(
+      2,
+      nLay + 1,
+      nLay));
+  Modelica.Blocks.Interfaces.RealInput solDir
+    "direct solar illuminance on surface se"
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: what does 'se' mean?

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    annotation (Placement(transformation(extent={{-120,40},{-80,80}})));
+  Modelica.Blocks.Interfaces.RealInput solDif
+    "diffuse solar illuminance on surface s"
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: what does 's' mean?

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+    annotation (Placement(transformation(extent={{-120,0},{-80,40}})));
+  Modelica.Blocks.Interfaces.RealInput angInc "angle of incidence"
+    annotation (Placement(transformation(extent={{-120,-80},{-80,-40}})));
+  Modelica.Thermal.HeatTransfer.Interfaces.HeatPort_a[nLay] iSolAbs
+    "solar absorptance in the panes"
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: technically, it's a heat flow rate, not an absorptance (which I interpret as a coefficient)

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    annotation (Placement(transformation(extent={{-10,90},{10,110}})));
+  Modelica.Thermal.HeatTransfer.Interfaces.HeatPort_a iSolDir
+    "transmitted direct solar riadtion"
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: 'radiation'

File: IDEAS/LIDEAS/Components/BaseClasses/LinearAirModel.mo [hunk 8]
```diff
@@ -0,0 +1,87 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+model LinearAirModel
+  extends IDEAS.Buildings.Components.ZoneAirModels.BaseClasses.PartialAirModel(
+    TAir(stateSelect=if sim.linearise then StateSelect.prefer else StateSelect.default),
+    useFluPor=not linearise and useFluPor2);
+  parameter Boolean useFluPor2 = true "Set to false to remove fluid ports, overwritten by sim.linearise";
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinearAirModel.mo]: do we still need this functionality? Otherwise best to remove it and first merge #837

File: IDEAS/LIDEAS/Components/BaseClasses/LinearAirModel.mo [hunk 9]
```diff
@@ -0,0 +1,87 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+model LinearAirModel
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/LinearAirModel.mo]: I don't think we will still need a `LinearAirModel` since the existing `WellMixedAir` can be used directly?

File: IDEAS/LIDEAS/Components/BaseClasses/Outputs.mo [hunk 10]
```diff
@@ -0,0 +1,6 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+expandable connector Outputs
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/Outputs.mo]: Do we need `expandable` here?

File: IDEAS/LIDEAS/Components/BaseClasses/OutputsWithStates.mo [hunk 11]
```diff
@@ -0,0 +1,6 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+expandable connector OutputsWithStates
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/OutputsWithStates.mo]: Do we need `expandable` here?

File: IDEAS/LIDEAS/Components/BaseClasses/Prescribed.mo [hunk 12]
```diff
@@ -0,0 +1,7 @@
+within IDEAS.LIDEAS.Components.BaseClasses;
+expandable connector Prescribed
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/BaseClasses/Prescribed.mo]: Do we need `expandable` here?

File: IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo [hunk 13]
```diff
@@ -0,0 +1,16 @@
+within IDEAS.LIDEAS.Components;
+model LinRectangularZoneTemplate
+  extends IDEAS.Buildings.Components.RectangularZoneTemplate(redeclare replaceable
+      BaseClasses.LinearAirModel airModel(linearise=linearise),
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo]: I think we can use the `WellMixedAir` model here when it includes `T(stateSelect=StateSelect.prefer)`?

File: IDEAS/LIDEAS/Components/LinWindow.mo [hunk 14]
```diff
@@ -0,0 +1,33 @@
+within IDEAS.LIDEAS.Components;
+model LinWindow
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/LinWindow.mo]: Should have a comment explaining the model.

File: IDEAS/LIDEAS/Components/LinZone.mo [hunk 15]
```diff
@@ -0,0 +1,6 @@
+within IDEAS.LIDEAS.Components;
+model LinZone "Linearisable zone model"
+  extends IDEAS.Buildings.Components.Zone(redeclare replaceable BaseClasses.LinearAirModel
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/LinZone.mo]: I think that we can used `WellMixedAir` directly.

Comment (Mathadon) [reply IDEAS/LIDEAS/Components/LinZone.mo]: And it looks like that we will not need a `LinZone` model anymore then?

File: IDEAS/LIDEAS/Examples/BaseClasses/SimInfoManagerInputs.mo [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+  connect(realExpression.y, localWeather.u) annotation (Line(points={{-63,-10},
+          {-58,-10},{-54,-10}},                     color={0,0,127}));
+  annotation (Diagram(coordinateSystem(preserveAspectRatio=false, extent={{-100,
+            -100},{100,100}})), Documentation(info="<html>
+<p>This extension of the SimInfoManager allows to use a custom text input file for the weather data instead of the TMY3 weather data loaded by the SimInfoManager.</p>
+<p>The text file should have the following structure. It should start with:</p>
+<p>double data(xx,10)</p>
+<p>where xx should be replaced by the number of time entries (number of rows).</p>
+<p>The following lines should have 10 tab-separated columns containing following values:</p>
+<p>time [s], ambiant temperature [K], dew point temperature [K], atmospheric pressure [bar], wind speed [m/s], wind bearing [deg] (between 0 and 360), relative humidity [-] (between 0 and 1), direct normal sun irradiance [W/m2], diffuse horizontal irradiance [W/m2] and weather the measurement is trusted or not (even though nothing is done in modelica with the last column).</p>
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Examples/BaseClasses/SimInfoManagerInputs.mo]: As per IBPSA convention, please break such long lines into multiple parts.

File: IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationLinear.mos [hunk 17]
```diff
@@ -0,0 +1,27 @@
+removePlots();
```
Comment (Mathadon) [inline IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationLinear.mos]: Please add reference results for this unit test.

File: IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationLinearInputs.mos [hunk 18]
```diff
@@ -0,0 +1,27 @@
+removePlots();
```
Comment (Mathadon) [inline IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationLinearInputs.mos]: Please add reference results for this unit test.

File: IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationNonLinear.mos [hunk 19]
```diff
@@ -0,0 +1,27 @@
+removePlots();
```
Comment (Mathadon) [inline IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationNonLinear.mos]: Please add reference results for this unit test.

File: IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationNonLinearInputs.mos [hunk 20]
```diff
@@ -0,0 +1,27 @@
+removePlots();
```
Comment (Mathadon) [inline IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationNonLinearInputs.mos]: Please add reference results for this unit test.

File: IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationLinear.mos [hunk 21]
```diff
@@ -0,0 +1,26 @@
+removePlots();
```
Comment (Mathadon) [inline IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationLinear.mos]: Please add reference results for this unit test.

File: IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationNonLinear.mos [hunk 22]
```diff
@@ -0,0 +1,26 @@
+removePlots();
+// --------------- Linearization and save matrices into ssm.mat --------------------------------
```
Comment (Mathadon) [inline IDEAS/Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationNonLinear.mos]: Please add reference results for this unit test.

File: IDEAS/LIDEAS/Components/BaseClasses/Inputs.mo (hunk 2 above)
Comment (damienpicard) [reply IDEAS/LIDEAS/Components/BaseClasses/Inputs.mo]: Done.

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo (hunk 3 above)
Comment (damienpicard) [reply IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: done

**Comment by damienpicard** [comment]
@Mathadon I still get a number of warnings. Are those problematics?

- For  IDEAS.LIDEAS.Validation.Case900Linearise
```
Modifier for protected component vol. This is not legal Modelica.
File: C:/Users/u0085306/Documents/work/modelica/IDEAS/IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo, line 5
Context: IDEAS.LIDEAS.Components.LinRectangularZoneTemplate.IDEAS.Buildings.Components.RectangularZoneTemplate.airModel.vol
Original declaration of vol:
File: C:/Users/u0085306/Documents/work/modelica/IDEAS/IDEAS/Buildings/Components/ZoneAirModels/WellMixedAir.mo, line 10
Context: IDEAS.Buildings.Components.ZoneAirModels.WellMixedAir.vol

Redeclaration requires a subtype.
But missing public function saturationPressure.
In redeclaration of Medium.
File: C:/Users/u0085306/Documents/work/modelica/IDEAS/IDEAS/Buildings/Components/InterzonalAirFlow/BaseClasses/PartialInterzonalAirFlow.mo, line 3
Context: IDEAS.Buildings.Components.Interfaces.PartialZone.Medium

The input winBusIn[1].nLay had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.numSolBus had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.solBus[1].outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.solBus[2].outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.solBus[3].outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.solBus[4].outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.solBus[5].outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

The input weaBus.solBus[6].outputAngles had a binding equation.
It will not be treated as an input - use a literal start-attribute if you want to keep it as input and provide a default value (an input is intended to get its value externally).

Inconsistency for connectors dimensioned using parameter with annotation(Dialog(connectorSizing=true)).
Some connector elements are not connected.
This is either because the connections and/or parameters has been edited textually to an inconsistent state
or because the default for the parameter is not 0. The unconnected connectors are:
linRecZon.airModel.vol.ports[3]
```

- for `IDEAS.LIDEAS.Validation.ZoneWithInputsValidationNonLinear` I also have a bunch of warnings. One of the problem is that I read all state variables (also the protected ones) and I used them to initialize the state space model

**Comment by Mathadon** [comment]
@damienpicard can you merge the master and then fix the failing unit tests before I review this?
Let's ignore the warnings for now if they only appear in LIDEAS.

**Comment by Mathadon** [comment]
@damienpicard it seems like all unit tests fail
```
*** Error: Failed to read /tmp/tmp-IDEAS-0-_9Sl0u/IDEAS/Case900ValidationNonLinear.mat generated by LIDEAS/Validation/Case900ValidationNonLinear.mos.
[Errno 2] No such file or directory: '/tmp/tmp-IDEAS-0-_9Sl0u/IDEAS/Case900ValidationNonLinear.mat'
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationNonLinear'.
*** Error: Simulation failed for 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationNonLinear.mos");'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.ZoneWithInputsValidationNonLinear'.
*** Error: Simulation failed for 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationNonLinear.mos");'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationNonLinearInputs'.
*** Error: Simulation failed for 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationNonLinearInputs.mos");'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationLinearInputs'.
*** Error: Simulation failed for 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationLinearInputs.mos");'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationLinear'.
*** Error: Simulation failed for 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/Case900ValidationLinear.mos");'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.ZoneWithInputsValidationLinear'.
*** Error: Simulation failed for 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationLinear.mos");'.
Number of models that failed check                           : 6
Number of models that failed to simulate                     : 6
Script that runs unit tests had 0 warnings and 13 errors.
```
perhaps you forgot to include one of the models in git?

**Comment by Mathadon** [comment]
I'm getting this warning:

```
The package.order is incomplete, since the class HeatTransfer in file/directory /media/psf/Home/Documents/git/IDEAS/IDEAS/HeatTransfer.mo is missing.
``` 

The unit tests fail because `IDEAS.Buildings.Components.Shading.ShadingControl` uses `irr`, which has been removed in the past. Apparently `ShadingControl` was not used in any unit test before due to which this went unnoticed.

**Comment by Mathadon** [comment]
@damienpicard I fixed the `irr` issue in d89d22d0ae5e4ee1409735ea491ce3aab50bc2a2. However, the problem is now singular. Quite likely due to changes in the `SimInfoManager` (addition of `WeaDatBus` etc). Can you have a look at this?

**Comment by Mathadon** [comment]
Perhaps it's also best to do a regular redeclare (or redeclare final) instead of `redeclare replaceable Buildings.Components.ZoneAirModels.WellMixedAir`. Otherwise people may accidentally undo the changes in `LinZone`.

**Comment by Mathadon** [comment]
Fixed the singularity.

Now getting warnings in unit tests:
```
*** Warning: Unspecified initial conditions in 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationLinear.mos");'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationNonLinearInputs'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationNonLinear'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationLinearInputs'.
*** Error: Model check failed for 'IDEAS.LIDEAS.Validation.Case900ValidationLinear'.
Number of models that failed check                           : 4
Number of models with unspecified initial conditions         : 1
```
despite that unit tests produce correct results.

For `ZoneWithInputsValidationLinear.mos`
```
The initial conditions are not fully specified.
Dymola has selected default initial conditions.

Assuming fixed start value for the continuous states:
stateSpace.x[5](start = stateSpace.x_start[5])
stateSpace.x[6](start = stateSpace.x_start[6])
```
despite that initial equations do exist..

**Comment by Mathadon** [comment]
Todo:
 - [ ] Fix documentation
 - [ ] Revise code
 - [ ] Review results
 - [ ] Check taco

**Comment by Mathadon** [comment]
Unit tests fail since
`
final parameter Integer nInp = Bsize[2] "Number of inputs";
`
is a structural parameter that depends on 
`final parameter Integer[2] Bsize = readMatrixSize(fileName=fileName, matrixName="B") "Size of B matrix of state space model";`
If that matrix does not (yet) exist, the parameter cannot be read. So dymola complains when checking the model. The matrix sizes should be defined in a different way.

File: IDEAS/Buildings/Components/RectangularZoneTemplate.mo [hunk 23]
```diff
@@ -557,12 +629,47 @@ protected
...
+  shaCorr=shaTypC.shaCorr)),
+    fraType(present=fraTypC.present, U_value=fraTypC.U_value),
+    linExtRad=linExtRadWin)
     "Window for face C of this zone" annotation (Placement(
         transformation(
         extent={{-5,-10},{5,10}},
         rotation=0,
         origin={-95,-30})));
+  replaceable
   IDEAS.Buildings.Components.Window winD(inc=IDEAS.Types.Tilt.Wall, azi=aziA +
```
Comment (Mathadon) [inline IDEAS/Buildings/Components/RectangularZoneTemplate.mo]: https://github.com/open-ideas/IDEAS/pull/892 has to be merged first

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo (hunk 4 above)
Comment (Mathadon) [reply IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: fixed

File: IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo (hunk 5 above)
Comment (Mathadon) [reply IDEAS/LIDEAS/Components/BaseClasses/LinWindowResponse.mo]: fixed

File: IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+    redeclare LinWindow winA(indexWindow=indexWindowA),
+    redeclare LinWindow winCei(indexWindow=indexWindowCei),
+    redeclare LinWindow winB(indexWindow=indexWindowB),
+    redeclare LinWindow winC(indexWindow=indexWindowC),
+    redeclare LinWindow winD(indexWindow=indexWindowD));
+  parameter Boolean linearise=sim.linearise;
+  parameter Integer indexWindowA=1 if hasWinA "Index of this window A";
+  parameter Integer indexWindowB=1 if hasWinB "Index of this window B";
+  parameter Integer indexWindowC=1 if hasWinC "Index of this window C";
+  parameter Integer indexWindowD=1 if hasWinD "Index of this window D";
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo]: is this correct?

**Comment by damienpicard** [comment]
@Mathadon Thanks for looking at all that.

Regarding your question:
- _This is fixed if I remove commonWall(layMul(monLay(each monLayDyn(addRes_b=true)) in IDEAS.LIDEAS.Validation.ZoneWithInputsValidationLinear. @damienpicard why did we add this line? Can it be removed?_ --> if I remember correcly this was necessary to have the exact same model for the linearized model and the SSM. I'm not sure about the details anymore.

- _Unit tests fail since
final parameter Integer nInp = Bsize[2] "Number of inputs";
is a structural parameter that depends on
final parameter Integer[2] Bsize = readMatrixSize(fileName=fileName, matrixName="B") "Size of B matrix of state space model";
If that matrix does not (yet) exist, the parameter cannot be read. This is the case in the unit test script, which performs a 'check' before executing the .mos script. The unit tests therefore fail although the simulation results from the script are generated correctly since it generates the matrix input file._ --> as in the model the name of each state is hard-coded, we could as well hard-code the number of states. This is less automatic but it is not a problem.

File: IDEAS/LIDEAS/Examples/ZoneLinearise.mo [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+    PWall=3,
+    inc=IDEAS.Types.Tilt.Floor,
+    azi=0) "Floor model"
+    annotation (Placement(transformation(extent={{-54,20},{-44,40}})));
+  IDEAS.Buildings.Components.OuterWall outerWall(
+    azi=0,
+    redeclare parameter IDEAS.Buildings.Validation.Data.Constructions.HeavyWall
+      constructionType,
+    A=10,
+    inc=1.5707963267949) "Outer wall model"
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Examples/ZoneLinearise.mo]: why not use IDEAS.Types.Tilt.Wall?

File: IDEAS/LIDEAS/Examples/ZoneLinearise.mo [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+    annotation (Placement(transformation(extent={{20,-70},{40,-50}})));
+  IDEAS.Buildings.Components.Shading.ShadingControl shadingControl
+    "Shading control model"
+    annotation (Placement(transformation(extent={{-80,-100},{-60,-80}})));
+  IDEAS.Buildings.Components.BoundaryWall commonWall1(
+    redeclare parameter IDEAS.Buildings.Validation.Data.Constructions.HeavyWall
+      constructionType,
+    A=10,
+    azi=0,
+    inc=1.5707963267949,
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Examples/ZoneLinearise.mo]: why not use IDEAS.Types.Tilt.Wall?

File: IDEAS/LIDEAS/Examples/ZoneLinearise.mo [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+    nSurf=4) "first floor"
+    annotation (Placement(transformation(extent={{20,-70},{40,-50}})));
+  IDEAS.Buildings.Components.Shading.ShadingControl shadingControl
+    "Shading control model"
+    annotation (Placement(transformation(extent={{-80,-100},{-60,-80}})));
+  IDEAS.Buildings.Components.BoundaryWall commonWall1(
+    redeclare parameter IDEAS.Buildings.Validation.Data.Constructions.HeavyWall
+      constructionType,
+    A=10,
+    azi=0,
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Examples/ZoneLinearise.mo]: why not use why not use IDEAS.Types.Azimuth.S?

File: IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo (hunk 24 above)
Comment (damienpicard) [reply IDEAS/LIDEAS/Components/LinRectangularZoneTemplate.mo]: The value of the index cannot be guessed upfront as it depends also on the number of zone. It is probably better not to give any default value to it.

File: IDEAS/LIDEAS/Examples/ZoneLinearise.mo [hunk 28]
```diff
@@ -0,0 +1,187 @@
...
+  connect(floor.propsBus_a, firFlo.propsBus[1]) annotation (Line(
+      points={{7,-42.1667},{7,-54.5},{20,-54.5}},
+      color={255,204,51},
+      thickness=0.5));
+  annotation (Diagram(coordinateSystem(preserveAspectRatio=false, extent={{-100,
+            -100},{100,100}})),           __Dymola_Commands(file=
+          "Scripts/linearize_ZoneLinearise.mos" "Linearise"),
+    Documentation(revisions="<html>
+<ul>
+<li>August 21, 2018 by Damien Picard: <br/>Change medium to <code>IDEAS.Media.Specialized.DryAir</code>.</li>
```
Comment (Mathadon) [inline IDEAS/LIDEAS/Examples/ZoneLinearise.mo]: Can you spread the revision history across multiple lines in the future like the IBPSA convention?
```
<li>
August 21, 2018 by Damien Picard: <br/>
Change medium to <code>IDEAS.Media.Specialized.DryAir</code>.
</li>
```

File: IDEAS/Buildings/Components/RectangularZoneTemplate.mo (hunk 23 above)
Comment (damienpicard) [reply IDEAS/Buildings/Components/RectangularZoneTemplate.mo]: This is problematic has we want to by able to redeclare the windows but the way it is coded now has  a constrainedby with all variables filled in. The solution is to make a partial rectangularZoneTemplate which contains everything except the windows. The LinRectangularZoneTemplate with then use directly the LinWindow (no need for redeclaration).

**Comment by Mathadon** [comment]
Apparently this does cause unit tests to fail:

```
*** Warning: Unspecified initial conditions in 'RunScript("Resources/Scripts/Dymola/LIDEAS/Validation/ZoneWithInputsValidationLinear.mos");'.
```

**Comment by Mathadon** [comment]
@damienpicard the initialization issue is fixed in 07ebc6e. The problem wasn't that the state space initial equation was missing, but that there was no initial equation for the actual state, since `EnergyDynamics=DynamicFreeInitial` when connecting to a fixed temperature boundary condition.

**Comment by Mathadon** [comment]
This is ready to be merged when the unit tests finish.
